src/ingest/src/main/scala/osmesa/ingest/util/Caching.scala

```diff
@@ -0,0 +1,71 @@
+package osmesa.ingest.util
+
+import com.amazonaws.services.s3.AmazonS3ClientBuilder
+import org.apache.spark.sql._
+import org.apache.log4j.{Level, Logger}
+
+import java.io.File
+
+
+trait Caching {
+  def orc(filename: String)(sparkjob: => DataFrame)(implicit ss: SparkSession): DataFrame
+}
+
+class S3Caching(cacheDir: String) extends Caching {
+  lazy val logger = Logger.getRootLogger()
+
+  private val s3client = AmazonS3ClientBuilder.defaultClient()
+
+  private val bucket = cacheDir.split("/")(2)
+
+  private def prefix(filename: String): String =
+    cacheDir.split("/").drop(3).mkString("/") + "/" + filename
+
+  private def fileUri(filename: String): String =
+    cacheDir + "/" + filename
+
+  private def s3exists(filename: String): Boolean =
+    s3client.listObjectsV2(bucket, prefix(filename)).getKeyCount > 0
+
+  def orc(filename: String)(sparkjob: => DataFrame)(implicit ss: SparkSession): DataFrame = {
+    logger.warn("in cache func")
+    logger.warn("bucket: " ++ bucket)
+    logger.warn(prefix(filename))
+    logger.warn(fileUri(filename))
+    logger.warn("s3exists: " ++ s3exists(filename).toString)
+    if (s3exists(filename)) {
+      ss.read.orc(fileUri(filename)).repartition(10000)
+    } else {
+      sparkjob.repartition(1).write.format("orc").save(fileUri(filename))
```

Comment:
If you mean the `repartition(1)`, I fully agree.  ORC writes were taking 20+ minutes for North America.  Given the number of times we cache, this adds up in a significant way.

Reply:
Eventually (and generally), it would be nice to figure out a heuristic for determining the number of target partitions. Too few and writes take forever. Too many and initializing a remote datasource takes a long time (but only on the driver, and only when doing the initial "read"). I've been using an upper limit of 100, fwiw.

Reply:
Yes.  This.  These repartitions are a bad idea.  I surely did encounter issues with these configs.  I ended up eliminating the repartition on 37, and changing the repartition on 39 to be the same as the number of worker nodes.  Cleared up many problems.  Misunderstanding the degree of participation of the master node in coordinating tasks led to some mysterious behaviors.  Getting fast performance out of ORC would seem like an easy thing (considering that the phrase "High performance" is one of the first things you see when you search for `Apache ORC`), but it's clearly non-trivial.  I think ORC writes of planet-scale cache data were taking about 20 minutes each?  I've no context for knowing if that's unreasonable.

Reply:
¯\_(ツ)_/¯

It seems moderately reasonable for ~300GB of writes to S3, but the ability to parallelize is nice. More partitions possibly means more splits when reading (but also more metadata requests when initializing).

I'm really looking forward to seeing if the ORC improvements (`spark.sql.orc.impl=native`) in Spark 2.3.0 speed things up...

Using instance types that support the full 25Gbps _may_ help now that S3 supports those speeds as well.